src/cpp/client.cpp

```diff
@@ -30,386 +30,408 @@
-    this->_append_dataset_tensor_commands(cmds, dataset);
-    this->_append_dataset_ack_command(cmds, dataset);
-    this->_run(cmds);
-    return;
+    _append_dataset_metadata_commands(cmds, dataset);
+    _append_dataset_tensor_commands(cmds, dataset);
+    _append_dataset_ack_command(cmds, dataset);
+    _run(cmds);
 }
 
+// Retrieve the current dataset
 DataSet Client::get_dataset(const std::string& name)
 {
     // Get the metadata message and construct DataSet
-    CommandReply reply = this->_get_dataset_metadata(name);
-
-    if(reply.n_elements()==0)
+    CommandReply reply = _get_dataset_metadata(name);
+    if (reply.n_elements() == 0) {
         throw std::runtime_error("The requested DataSet " +
                                  name + " does not exist.");
+    }
 
-    DataSet dataset = DataSet(name);
-    this->_unpack_dataset_metadata(dataset, reply);
+    DataSet dataset(name);
+    _unpack_dataset_metadata(dataset, reply);
 
     std::vector<std::string> tensor_names = dataset.get_tensor_names();
 
-    std::vector<size_t> reply_dims;
-    std::string_view blob;
-    TensorType type;
-    std::string tensor_key;
-    for(size_t i=0; i<tensor_names.size(); i++) {
-        tensor_key =
-            this->_build_dataset_tensor_key(name, tensor_names[i],
-                                            true);
+    for(size_t i = 0; i < tensor_names.size(); i++) {
+        std::string tensor_key =
+            _build_dataset_tensor_key(name, tensor_names[i], true);
         CommandReply reply = this->_redis_server->get_tensor(tensor_key);
-        reply_dims = GetTensorCommand::get_dims(reply);
-        blob = GetTensorCommand::get_data_blob(reply);
-        type = GetTensorCommand::get_data_type(reply);
+        std::vector<size_t> reply_dims = GetTensorCommand::get_dims(reply);
+        std::string_view blob = GetTensorCommand::get_data_blob(reply);
+        TensorType type = GetTensorCommand::get_data_type(reply);
         dataset._add_to_tensorpack(tensor_names[i],
                                    (void*)blob.data(), reply_dims,
                                    type, MemoryLayout::contiguous);
     }
+
+    // ***WS*** FINDME: Passing objects by value causes unnecessary copies
     return dataset;
 }
 
+// Rename the current dataset
 void Client::rename_dataset(const std::string& name,
                             const std::string& new_name)
 {
-    this->copy_dataset(name, new_name);
-    this->delete_dataset(name);
-    return;
+    copy_dataset(name, new_name);
+    delete_dataset(name);
 }
 
+// Clone the dataset to a new name
 void Client::copy_dataset(const std::string& src_name,
                           const std::string& dest_name)
 {
-    CommandReply reply = this->_get_dataset_metadata(src_name);
-    DataSet dataset = DataSet(src_name);
-    this->_unpack_dataset_metadata(dataset, reply);
+    // Extract metadata
+    CommandReply reply = _get_dataset_metadata(src_name);
+    if (reply.n_elements() == 0) {
```

Comment:
Maybe this check would be better in ``_unpack_dataset_metadata`` so it gets used every time?